File: src/advanced/kmp.cpp
// KMP 字串比對（CLRS §32.4）：前綴函數 π ＋ O(n) 掃描
#include <iostream>
#include <string>
#include <vector>

std::vector<int> prefixFunction(const std::string& P) {
    int m = P.size();
    std::vector<int> pi(m + 1, 0);                 // 1-indexed：pi[q] = 最長真前綴長
    int k = 0;
    for (int q = 2; q <= m; ++q) {
        while (k > 0 && P[k] != P[q - 1]) k = pi[k];
        if (P[k] == P[q - 1]) ++k;
        pi[q] = k;
    }
    return pi;
}

int main() {
    std::string T = "abababacaba", P = "ababaca";
    auto pi = prefixFunction(P);
    std::cout << "π(ababaca) = [";
    for (size_t q = 1; q < pi.size(); ++q) std::cout << pi[q] << (q + 1 < pi.size() ? " " : "]\n");

    int n = T.size(), m = P.size(), q = 0;         // q = 已匹配的字元數
    for (int i = 0; i < n; ++i) {
        while (q > 0 && P[q] != T[i]) q = pi[q];   // 失配：退到 π[q]，不回頭讀 T
        if (P[q] == T[i]) ++q;
        if (q == m) {
            std::cout << "在 shift " << i - m + 1 << " 找到 pattern\n";
            q = pi[q];
        }
    }
    return 0;
}

File: src/advanced/kmp.py
# KMP（改寫自官方 Chapter 32/kmp_matcher.py，MIT License）：1-based π 表版
def prefix_function(P):
    m = len(P)
    pi = [0] * (m + 1)                       # pi[q] = P[:q] 的最長真前綴兼後綴長
    k = 0
    for q in range(2, m + 1):
        while k > 0 and P[k] != P[q - 1]:
            k = pi[k]
        if P[k] == P[q - 1]:
            k += 1
        pi[q] = k
    return pi


def kmp_matcher(T, P):
    pi = prefix_function(P)
    q = 0                                    # 已匹配的字元數
    hits = []
    for i, ch in enumerate(T):
        while q > 0 and P[q] != ch:          # 失配：q 退到 π[q]，i 不回頭
            q = pi[q]
        if P[q] == ch:
            q += 1
        if q == len(P):
            hits.append(i - len(P) + 1)
            q = pi[q]
    return pi, hits


P = "ababaca"
pi, hits = kmp_matcher("abababacaba", P)
print("π 表（q=1..7）：")
for q in range(1, len(P) + 1):
    print("  q=%d  P[:q]=%-8s π[q]=%d" % (q, P[:q], pi[q]))
print("在 shift", hits, "找到 pattern")
